packages/flutter_tools/lib/src/web/devfs_web.dart

```diff
@@ -131,30 +163,186 @@ class WebAssetServer {
+    bool bundleFirstUpload = false,
+    @required ResidentCompiler generator,
+    String dillOutputPath,
+    @required bool trackWidgetCreation,
+    bool fullRestart = false,
+    String projectRootPath,
+    String pathToReload,
+    List<Uri> invalidatedFiles,
+  }) async {
+    assert(trackWidgetCreation != null);
+    assert(generator != null);
+    if (bundleFirstUpload) {
+      final File requireJS = fs.file(fs.path.join(
+        artifacts.getArtifactPath(Artifact.engineDartSdkPath),
+        'lib',
+        'dev_compiler',
+        'kernel',
+        'amd',
+        'require.js',
+      ));
+      final File dartSdk = fs.file(fs.path.join(
+        artifacts.getArtifactPath(Artifact.flutterWebSdk),
+        'kernel',
+        'amd',
+        'dart_sdk.js',
+      ));
+      final File dartSdkSourcemap = fs.file(fs.path.join(
+        artifacts.getArtifactPath(Artifact.flutterWebSdk),
+        'kernel',
+        'amd',
+        'dart_sdk.js.map',
+      ));
+      final File stackTraceMapper = fs.file(fs.path.join(
+        artifacts.getArtifactPath(Artifact.engineDartSdkPath),
+        'lib',
+        'dev_compiler',
+        'web',
+        'dart_stack_trace_mapper.js',
+      ));
+      _webAssetServer.writeFile('/main.dart.js', generateBootstrapScript(
+        requireUrl: requireJS.path,
+        mapperUrl: stackTraceMapper.path,
+        entrypoint: '$mainPath.js',
+      ));
+      _webAssetServer.writeFile('/main_module.js', generateMainModule(
+        entrypoint: '$mainPath.js',
+      ));
+      _webAssetServer.writeFile('/dart_sdk.js', dartSdk.readAsStringSync());
+      _webAssetServer.writeFile('/dart_sdk.js.map', dartSdkSourcemap.readAsStringSync());
+    }
+    final DateTime candidateCompileTime = DateTime.now();
+    if (fullRestart) {
+      generator.reset();
+    }
+     final CompilerOutput compilerOutput = await generator.recompile(
+      mainPath,
+      invalidatedFiles,
+      outputPath:  dillOutputPath ?? getDefaultApplicationKernelPath(trackWidgetCreation: trackWidgetCreation),
+      packagesFilePath : _packagesFilePath,
+    );
+    if (compilerOutput == null || compilerOutput.errorCount > 0) {
+      return UpdateFSReport(success: false);
+    }
+    // Only update the last compiled time if we successfully compiled.
+    lastCompiled = candidateCompileTime;
+    // list of sources that needs to be monitored are in [compilerOutput.sources]
+    sources = compilerOutput.sources;
+    File codeFile;
+    File manifestFile;
+    File sourcemapFile;
+    List<String> modules;
+    try {
+      codeFile = fs.file('${compilerOutput.outputFilename}.sources');
+      manifestFile = fs.file('${compilerOutput.outputFilename}.json');
+      sourcemapFile = fs.file('${compilerOutput.outputFilename}.map');
+      modules = _webAssetServer.write(codeFile, manifestFile, sourcemapFile);
+    } on FileSystemException {
+      throwToolExit('Incremental compiler did not produce expected output.');
```

Comment:
nit: this message may imply that the files were successfully written but their contents are somehow wrong, while it seems to be a plain failure to write files.

Reply:
BTW, `FileSystemException` could contain some useful info such as `osError` and `path`.

Reply:
Updated the exit to include the original error message